### DISCUSSION TITLE
Need instructions on how to generate a custom ebook

### DISCUSSION THREAD

**Discussion by jz533** [DivinumOfficium/divinum-officium#3344] [Q&A]
Title says it all. How do I go about using the provided scripts and such to generate an ebook? In this case, I want to generate an ebook for Matins only (if possible), in the pre-1955 Office. If someone could get me on the right track (i.e. which file to use and how to run the script), I would be glad to figure out as much as possible from there.

Thanks!

**Comment by APMarcello3** [comment]
Cf. Parvum in epub/mobi #1087

**Comment by APMarcello3** [comment]
Cf. Can epubgen be used to create daily Mass propers? #1470

**Comment by Geremia** [comment]
@jz533 

Apply these changes:
```diff
diff --git a/standalone/tools/epubgen2/epubgen2.sh b/standalone/tools/epubgen2/epubgen2.sh
index 4d1c42faf..0696bdecb 100755
--- a/standalone/tools/epubgen2/epubgen2.sh
+++ b/standalone/tools/epubgen2/epubgen2.sh
@@ -68,10 +68,10 @@ ALL_RUBRICS=("Trident 1570" "Trident 1910" "Divino Afflatu" "Reduced 1955" "Rubr
 ALL_RUBRICS_NAME=("_1570" "_1910" "_DA" "_1955" "" "NC")
 
 MONTHNAMES=(Ianuarius Februarius Martius Aprilis Maius Iunius Iulius Augustus September October November December)
-HORA_INDEX_LAST=7
-HORAS_FILENAMES=(1-Matutinum 2-Laudes 3-Prima 4-Tertia 5-Sexta 6-Nona 7-Vespera 8-Completorium)
-HORAS_NAMES=(Matutinum Laudes Prima Tertia Sexta Nona Vespera Completorium)
-HORAS_NAMES_SHORT=(M L P T S N V C)
+HORA_INDEX_LAST=0
+HORAS_FILENAMES=(1-Matutinum)
+HORAS_NAMES=(Matutinum)
+HORAS_NAMES_SHORT=(M)
 
 YEAR_RE='^[0-9]+$'
```

And run using the "-r" option—e.g., for "Divino Afflatu":
`epubgen2.sh -r DA -o output_dir`

**Comment by mikolodz** [comment]
Hello, I'm using the up-to-date version with -m option and it's not working properly.

After the epub files are generated, inside them I can see the following error on every epub page:
<img width="455" alt="Zrzut ekranu 2020-08-30 o 16 56 18" src="https://user-images.githubusercontent.com/8095324/91662334-155ac480-eae2-11ea-92ad-f879b95730ba.png">

Anyone can help me out?
I'm using:
./epubgen2.sh -m

**Comment by APMarcello3** [comment]
Cf. issue #1794

**Comment by Geremia** [comment]
@mikolodz @APMarcello3 I created the code to add the Mass, but I can't figure out why it says "Post"; it should say "Mass".

**Comment by mikolodz** [comment]
@Geremia Thank you. Is your fork already merged into master? I've pulled the repo and tried to run the script during Christmas, but apparently, I must have updated some packages since August, and the script throws errors. It would be great to confirm if I can test it again using your fix.
BTW. I also don't know why does it say "Post". I wonder if I didn't mess something up when setting the epub export settings.

**Comment by Geremia** [comment]
@mikolodz Yes, it was merged.
Changing "Post" in EofficiumXhtml.pl:219 and EofficiumXhtml.pl:253 had no effect, so the "Post" is coming from elsewhere. I think it's related to the Ante / Post (prayers before and after the Office).

**Comment by jz533** [comment]
Curious, is there any way to create an ebook for the entire year of 2021 that is all in Latin, except for the lessons (and maybe responsories, too) of Matins in English? Trying to ease myself off of using English in the DO, but the lessons are still a must for now.

**Comment by APMarcello3** [comment]
@Geremia It would be very helpful to have some robust documentation on how to generate such files for the Office and the Mass, because a number of requests have come in recently on this matter.

**Comment by havenoskills** [comment]
Please forgive my ignorance of github and perl in advance. @jz533 asks if it's possible to do an entire year in Latin -- yes. But it requires commenting out a bunch of lines in the code. 

My purpose was slightly different: creating a Latin-only missal (you can easily keep all the office files by looking for the word 'skip' in my code and reverting the change). Since I don't actually understand perl, I had to preserve the html files and then post-process them with two python scripts. I'll merge the python scripts at some point. These are the net changes made by the process:

- Save all files as html for processing in Calibre (change the hardcoded directory as needed)
- Restored the red large initial capitals for each paragraph. I like the look.
- The TOC python script creates a one-year table of contents in the form of a calendar with the holy days of obligation in bold (one version was made for the original code that produced 3 entries for midnight/dawn/day Christmas)
- Eliminate the "ante" "communio" and "post" links that don't work in an ebook
- Added communion prayers (the original Laszlo Kiss code had Anima Christe, I added St Thomas' prayers)
- At the end of each Sunday mass, hit "Next Sunday" for the next Sunday mass (or turn the page for the daily mass)

So the usage is, put in the new files in the locations noted by the folder and run epubgen4.sh instead of epubgen2.sh. It doesn't overwrite anything in the repository. Make the table of contents by running the toc-print script, then run the scrubtext script. Import the resulting files into Calibre to make the ebook.

I wish I knew perl well enough to do a diff properly and convert the changes into command line options, but since I don't maybe the zip files can help someone else. I just kept poking away until the code did what I wanted. This is an amazing repository.

[newfiles.zip](https://github.com/DivinumOfficium/divinum-officium/files/5835480/newfiles.zip)

**Comment by mikolodz** [comment]
@Geremia Hello after a while. I have tested out the newest epubgen2 version. With some tweaking (changing the date command syntax in the script) it generates the epub (on MacOs without any further config or using Ubuntu 18.04 after installing multiple packages), but the generated epub is not working properly on any device I can access.

I am able to open it up using Calibre with some problems with urls, but otherwise it's not possible either on PocketBook nor on iPad / iPhone etc. I'm just getting a message that the ebook contains errors. On PocketBook it's just shutting down after loading.

Anyone made it to work on any ebook reader and mobile devices?
Could you please share some snippet or a working version for a reference?

**Comment by mbab** [comment]
Please try https://mbab1.gitlab.io/divinumofficium.epub/

**Comment by mikolodz** [comment]
@mbab Thank you brother! Will try it out!

**Comment by jz533** [comment]
Thanks a bunch! Looking forward to trying it.On May 24, 2023, at 7:02 AM, Michael ***@***.***> wrote:﻿
@mbab Thank you brother! Will try it out!

**Comment by birdylanfern** [comment]
> Please try https://mbab1.gitlab.io/divinumofficium.epub/

How exactly do I use this? pardon me for I know nothing about coding...

**Comment by mbab** [comment]
> > Please try https://mbab1.gitlab.io/divinumofficium.epub/
> 
> How exactly do I use this? pardon me for I know nothing about coding...

There is not need to know about coding.
You need to install *docker* (You can find many tutorials how to do this) and then run command as described
https://mbab1.gitlab.io/divinumofficium.epub/

**Reply by birdylanfern** [reply]
I've tried using this, but it keeps saying that the doepub command is not found...

**Reply by mbab** [reply]
Which command produced this output?
`docker run` ?

**Reply by jz533** [reply]
So I got it to to work in a virtual Docker environment online at https://labs.play-with-docker.com/. It generates the file and everything, BUT there is no way to download the file that is generated in the virtual environment's file system. SO CLOSE.

https://mbab1.gitlab.io/divinumofficium.epub/ are the instructions I followed. I wanted to get Monastic Matins, Latin - English. I got it to generate the entire year by using the command
$ docker run --rm -v $(pwd):/tmp mbab1/doepub -r Mon --horas Matutinum -l En -f 12-03-2023 -t 11-30-2024

SO CLOSE. It literally generates the whole file, but I literally have no way to download it to my local machine :(

I would try Docker on my local machine, but I spent hours running into problems setting it up, which is what led me to using the virtual environment.

Any way you could generate a quick file with that command and share it?

**Reply by APMarcello3** [reply]
@benyanke

**Reply by mbab** [reply]
I found such large file impractical, be here You have it: https://ufile.io/ea4axf45
Link will expire in 30 days.

**Reply by jz533** [reply]
Hello! @mbab That was extremely helpful.

If it was not much trouble, would you, or someone else, be okay with generating monthly files (2025) for Polish (newer) for the whole 1960 DO + Holy Mass? Hate to waste anyone's time, so would be happy to pay you back with some sort of online platform. I simply don't have the capabilities to run docker it seems.

**Reply by jz533** [reply]
Was looking for February - December.

**Reply by mbab** [reply]
https://ufile.io/5r65zjd3
